hbase-asyncfs/src/main/java/org/apache/hadoop/hbase/io/asyncfs/monitor/StreamSlowMonitor.java

```diff
@@ -0,0 +1,219 @@
+    "hbase.regionserver.async.wal.datanode.slow.packet.speed.min.kbs";
+  private static final double DEFAULT_DATANODE_SLOW_PACKET_FLUSH_MIN_SPEED = 20; // 20KB/s
+
+  private final String name;
+  // this is a map of datanodeInfo->queued slow PacketAckData
+  private final LoadingCache<DatanodeInfo, Deque<PacketAckData>> datanodeSlowDataQueue;
+  private final ExcludeDatanodeManager excludeDatanodeManager;
+
+  private int minSlowDetectCount;
+  private long slowDataTtl;
+  private long slowPacketAckMs;
+  private double minPacketFlushSpeedKBs;
+  private long minLengthForSpeedCheck;
+
+  public StreamSlowMonitor(Configuration conf, String name,
+      ExcludeDatanodeManager excludeDatanodeManager) {
+    setConf(conf);
+    this.name = name;
+    this.excludeDatanodeManager = excludeDatanodeManager;
+    this.datanodeSlowDataQueue = CacheBuilder.newBuilder()
+      .maximumSize(conf.getInt(WAL_MAX_EXCLUDE_SLOW_DATANODE_COUNT_KEY,
+        DEFAULT_WAL_MAX_EXCLUDE_SLOW_DATANODE_COUNT))
+      .expireAfterWrite(conf.getLong(WAL_EXCLUDE_DATANODE_TTL_KEY,
+        DEFAULT_WAL_EXCLUDE_DATANODE_TTL), TimeUnit.HOURS)
+      .build(new CacheLoader<DatanodeInfo, Deque<PacketAckData>>() {
+        @Override
+        public Deque<PacketAckData> load(DatanodeInfo key) throws Exception {
+          return new ConcurrentLinkedDeque<>();
+        }
+      });
+    LOG.info("New stream slow monitor {}", this.name);
+  }
+
+  public static StreamSlowMonitor create(Configuration conf, String name) {
+    return new StreamSlowMonitor(conf, name, new ExcludeDatanodeManager(conf));
+  }
+
+  /**
+   * Check if the packet process time shows that the relevant datanode is a slow node.
+   * @param datanodeInfo the datanode that processed the packet
+   * @param packetDataLen the data length of the packet (in bytes)
+   * @param processTimeMs the process time (in ms) of the packet on the datanode,
+   * @param lastAckTimestamp the last acked timestamp of the packet on another datanode
+   * @param unfinished if the packet is unfinished flushed to the datanode replicas
+   */
+  public void checkProcessTimeAndSpeed(DatanodeInfo datanodeInfo, long packetDataLen,
+      long processTimeMs, long lastAckTimestamp, int unfinished) {
+    long current = EnvironmentEdgeManager.currentTime();
+    // A datanode is suspicious slow if it meets one of the following conditions
+    // (please see more details in HBASE-26347):
+    // 1. no matter what the length of packet is, the the packet process time (in ms) is
+    //    greater than the configured slowPacketAckMs. This condition means that all the long enough
+    //    process time should be considered as slow.
+    // 2. if the data length of the packet is more than 4MB bytes, and the rate of process bytes is
+    //    less than the configured process speed. For large packet, speed should be considered.
+    boolean slow = processTimeMs > slowPacketAckMs || (packetDataLen > minLengthForSpeedCheck
```

Comment:
The condition is incorrect?

It should be
```java
boolean slow = (packetDataLen  <= minLengthForSpeedCheck && processTimeMs > slowPacketAckMs) || (packetDataLen > minLengthForSpeedCheck
&& (double) packetDataLen / processTimeMs < minPacketFlushSpeedKBs);
```

I think?

Reply:
I thought though large packets will check speeds, the process time will be larger than the configured value in general.
I'll change the condition to make it clearly.